pkg/queue/queue.go

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright (C) 2024 IOTech Ltd
+//
+
+package queue
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/IOTechSystems/go-mod-edge-utils/pkg/bootstrap/container"
+	"github.com/IOTechSystems/go-mod-edge-utils/pkg/di"
+	"github.com/IOTechSystems/go-mod-edge-utils/pkg/errors"
+)
+
+const (
+	DefaultMaxQueueCount = 10000
+	DefaultRetryInterval = "10m"
+)
+
+type ResendFun func(context.Context, *di.Container, any) bool
+
+type retryQueue struct {
+	dic           *di.Container
+	ctx           context.Context
+	maxCount      int
+	retryInterval time.Duration
+	items         []any
+	lock          sync.Mutex
+}
+
+// NewRetryQueue is a factory method that returns an initialized RetryQueue struct.
+func NewRetryQueue(dic *di.Container, ctx context.Context, maxCount int, retryInterval string, fun ResendFun) RetryQueueInterface {
+	logger := container.LoggerFrom(dic.Get)
+
+	if maxCount == 0 {
+		maxCount = DefaultMaxQueueCount
+	}
+
+	if retryInterval == "" {
+		retryInterval = DefaultRetryInterval
+	}
+
+	interval, err := time.ParseDuration(retryInterval)
+	if err != nil {
+		logger.Warnf("Failed to parse RetryInterval '%s', set to default '%s', err: %v", retryInterval, DefaultRetryInterval, err)
+		interval, _ = time.ParseDuration(DefaultRetryInterval)
+	}
+
+	q := &retryQueue{
+		dic:           dic,
+		ctx:           ctx,
+		maxCount:      maxCount,
+		retryInterval: interval,
+		lock:          sync.Mutex{},
+	}
+
+	logger.Debugf("Start RetryQueue with MaxQueueCount '%v' and RetryInterval '%s'", maxCount, interval)
+	go q.resendLoop(fun)
+
+	return q
+}
+
+// Concurrent safe Enqueue method that adds an new item to the RetryQueue
+func (q *retryQueue) Enqueue(item any) errors.Error {
+	logger := container.LoggerFrom(q.dic.Get)
+	q.lock.Lock()
+	if len(q.items) >= q.maxCount {
+		return errors.NewBaseError(errors.KindLimitExceeded, "Exceeded queue max count, drop the item", nil, nil)
+	}
+
+	q.items = append(q.items, item)
+	logger.Debug("Add new item to the queue")
+	q.lock.Unlock()
+	return nil
+}
+
+// Concurrent safe Dequeue method that removes the first item from the items of the RetryQueue
+func (q *retryQueue) Dequeue() {
+	logger := container.LoggerFrom(q.dic.Get)
+	q.lock.Lock()
+	q.items = q.items[1:]
+	logger.Debug("Remove the first item from the queue")
+	q.lock.Unlock()
+}
+
+// resendLoop method that triggers the resend function at intervals
+func (q *retryQueue) resendLoop(fun ResendFun) {
+	logger := container.LoggerFrom(q.dic.Get)
+	for {
+		select {
+		case <-q.ctx.Done():
+			logger.Info("Exiting retry loop")
+			return
+		case <-time.After(q.retryInterval):
+			if len(q.items) > 0 {
+				for index, item := range q.items {
```

Comment:
the items cannot be retrieved from loop like this, because it will cause race condition if someone calls Enqueue or Dequeue.
Add a new func like poll here:
https://docs.oracle.com/javase/8/docs/api/java/util/Queue.html

The items read or write should always be locked